lib/core/test/unit/Cardano/Wallet/Api/ServerSpec.hs

```diff
@@ -0,0 +1,22 @@
+module Cardano.Wallet.Api.ServerSpec (spec) where
+
+import Prelude
+
+import Cardano.Wallet.Api.Server
+    ( Listen (..), withListeningSocket )
+import Network.Socket
+    ( SockAddr (..), getSocketName, tupleToHostAddress )
+import Test.Hspec
+    ( Spec, describe, it, shouldReturn )
+
+spec :: Spec
+spec = describe "API Server" $ do
+    it "listens on the local interface" $ do
+        withListeningSocket "127.0.0.1" ListenOnRandomPort $ \(port, socket) -> do
+            getSocketName socket `shouldReturn`
+                SockAddrInet (fromIntegral port) (tupleToHostAddress (0x7f, 0, 0, 1))
+
+    it "can listen on any interface" $ do
+        withListeningSocket "0.0.0.0" ListenOnRandomPort $ \(port, socket) -> do
+            getSocketName socket `shouldReturn`
+                SockAddrInet (fromIntegral port) 0
```

Comment:
If we give `withListeningSocket` a bad host, it currently throws hard with something like:

```
cardano-wallet-jormungandr: Network.Socket.getAddrInfo (called with preferred socket type/protocol: AddrInfo {addrFlags = [AI_PASSIVE,AI_ADDRCONFIG], addrFamily = AF_UNSPEC, addrSocketType = Stream, addrProtocol = 0, addrAddress = <assumed to be undefined>, addrCanonName = <assumed to be undefined>}, host name: Just "patate", service name: Just "8090"): does not exist (Name or service not known)
```

which is also what the CLI will spit and is definitely not the best user experience. I suggest we:

1/ Catch this error in `withListeningSocket` and make it a value on which we can pattern-match
2/ Have one or more "negative" test case here.
3/ In `Cardano.Wallet.{HttpBridge,Jormungandr}.hs`, log that `withListeningSocket` has exited with an error that is human friendly and explain how the host is completely bonkers

Reply:
Yes good idea, thanks.

I have also caught the exception which occurs when the user tries to listen on a privileged port but does not have permission. So we could relax the `>= 1024` restriction on port numbers and still have a reasonable error message.

Also there is a deadlock issue in `failWith`, which I have hacked around